documentation/proposals/Proposal - Axis Input Devices.md

```diff
@@ -0,0 +1,624 @@
+    public bool IsGroupInDeadzone(IAxisDevice device, int groupIndex, Deadzone deadzone);
+    
+    // Requires that the length of the deadzones span matches the number of axes in the group
+    public bool IsGroupInDeadzone(IAxisDevice device, int groupIndex, ReadOnlySpan<Deadzone> deadzones);
+}
+```
+
+# Device Registration
+Due to the complex set of constraints detailed above, it is important that a validation layer exists to ensure that higher-level code can trust the device is following important conventions and constraints. For each device that connects, it must first register to the `AxisDeviceRegistry` in order to be used as an `IAxisDevice`.
+
+Any interested party can then reference the `AxisDeviceRegistry` to retrieve all available devices. Notably, this is missing a way to alert users of devices that have been newly connected or disconnected. Feedback is requested on this front: is this necessary or should it be implemented elsewhere?
+
+```csharp
+public class AxisDeviceRegistry
+{
+    /// <summary>
+    /// Required before usage of an IAxisDevice - this is essentially a "validation layer" where we can check if the device
+    /// configuration is valid, e.g. "do this AxisGroup's axes have definitions that satisfy the axis group?"
+    /// Things like checking for axes with a rotational definition, Left/Right handedness, groups' axis counts, etc
+    /// Will populate the proper AxisDeviceRegistry collections using runtime type checks
+    /// Registration is required to use a device each time it connects
+    ///
+    /// Should we develop a way to register these conditions at compile time as like a list of functions?
+    /// </summary>
+    public bool TryRegister(IAxisDevice device, [NotNullWhen(false)] out string? error);
+
+    /// <summary>
+    /// Unregisters a device
+    /// Will remove the device from the registry, required to be called when a device disconnects
+    /// </summary>
+    /// <param name="device"></param>
+    /// <returns>True if the device was previously registered and is now unregistered</returns>
+    public bool Unregister(IAxisDevice device);
+
+    /// <summary>
+    /// All currently registered devices
+    /// </summary>
+    public IReadOnlyList<IAxisDevice> Devices { get; }
+
+    /// <summary>
+    /// Runtime-generated virtual devices
+    /// </summary>
+    public IReadOnlyList<IAxisDevice> VirtualDevices { get; }
+
+    /// <summary>
+    /// Physical devices
+    /// </summary>
+    public IReadOnlyList<IAxisDevice> PhysicalDevices { get; }
+
+    // A future composability proposal can include additional properties to retrieve all available axis groups, irrespective
+    // of the source devices
+    
+    // there may also be a case for making the above 3 collection types (All, Virtual, Physical) a model that can be followed by 
+    // the standard input context, which would be the primary benefit to adding `IsVirtual` to `IInputDevice` - to be discussed in a future proposal
+}
+```

```

Comment:
I don't see why we need this when we have `InputContext` which could be changed to implicitly validate when it encounters an `IAxisDevice` in a backends, and connections/disconnections are handled in the obvious way as already specified by the base proposal. If there is a desire to register/deregister individual devices then there should be an implementation of `IInputBackend` that accommodates this.

Reply:
I would much prefer that - just wanted to silo things so as not to disturb The Great Base Proposal. Though, if we were to do that, I would also request the addition of the `VirtualDevices` and `Physical Devices` collections to the `InputContext`

Reply:
Is there really a case where just looping and testing whether IsVirtual is set isn’t sufficient? We can provide utility properties like these but I’d imagine that’d be all it’s doing under-the-hood given that I don’t think we’d want to allocate extra for something that probably isn’t used in hot paths (and if they are to be used in hot paths, then I’d encourage those users to handle the connection event and maintain their own lists)

Reply:
I think it is sufficient and isnt too much to ask, though i can see the case where *only* virtual or *only* physical devices will be preferred, and indexing could be hampered without dedicated collections

Reply:
Then the user can make those collections in that wrapper :)

I don't want to force everyone into overhead for those small use cases, and generally want to try and avoid having too many points-of-entry here as this could be more confusing.

Reply:
that is Valid and Agreeable!